### PR TITLE
Add SIP7 substandard 4 implementation

### PR DESCRIPTION
**Pull request by frankisawesome**
## Motivation
https://github.com/ProjectOpenSea/SIPs/blob/main/SIPS/sip-7.md
Add additional substandard 4 checks to the zone for bulk order hash verification

### CODE REVIEW

File: contracts/zones/ImmutableSignedZone.sol [hunk 1]
```diff
@@ -416,12 +396,84 @@ contract ImmutableSignedZone is
...
+        if (expectedConsiderationHash != actualConsiderationHash) {
+            revert SubstandardViolation(
+                3,
+                "invalid consideration hash",
+                orderHash
+            );
+        }
+
+        // substandard 4 - validate order hashes actual match expected
+        // context must have at least one orderHashes entry and be a multiple of 32 bytes
```
Comment (lfportal) [inline contracts/zones/ImmutableSignedZone.sol:436]: Strictly speaking `context` may have no order hash entries for substandard 4.

File: contracts/zones/ImmutableSignedZone.sol [hunk 2]
```diff
@@ -416,12 +396,84 @@ contract ImmutableSignedZone is
...
+        if (context.length % 32 != 0) {
+            revert InvalidExtraData(
+                "invalid context, expecting consideration hash followed by order hashes",
+                orderHash
+            );
+        }
+
+        bytes calldata orderHashesBytes = context[32:];
```
Comment (lfportal) [inline contracts/zones/ImmutableSignedZone.sol:444]: **Suggestion:** I think semantically it'd make more sense to perform the length and modulus check on `orderHashesBytes`.

File: contracts/zones/ImmutableSignedZone.sol [hunk 3]
```diff
@@ -480,4 +532,52 @@ contract ImmutableSignedZone is
...
+
+        // All elements from array1 exist in array2
+        return true;
+    }
+
+    /**
+     * @dev helper function to sort a bytes32 array
+     * @param arr input array
+     */
+    function _sortBytes32Array(bytes32[] memory arr) internal pure {
```
Comment (lfportal) [inline contracts/zones/ImmutableSignedZone.sol:570]: Looks like this function is unused.

**Review by lfportal — CHANGES_REQUESTED** [review]

File: contracts/zones/ImmutableSignedZone.sol (hunk 3 above)
Comment (frankisawesome) [reply contracts/zones/ImmutableSignedZone.sol]: [7487f3c](https://github.com/immutable/seaport-1.4/pull/30/commits/7487f3ceff21a4ffb06d39f41efdc4b8db03fd16)

File: contracts/zones/ImmutableSignedZone.sol (hunk 2 above)
Comment (frankisawesome) [reply contracts/zones/ImmutableSignedZone.sol]: [7487f3c](https://github.com/immutable/seaport-1.4/pull/30/commits/7487f3ceff21a4ffb06d39f41efdc4b8db03fd16)

File: contracts/zones/ImmutableSignedZone.sol (hunk 1 above)
Comment (frankisawesome) [reply contracts/zones/ImmutableSignedZone.sol]: [7487f3c](https://github.com/immutable/seaport-1.4/pull/30/commits/7487f3ceff21a4ffb06d39f41efdc4b8db03fd16)

**Review by lfportal — APPROVED** [review]

File: test/immutablesignedzone.spec.ts [hunk 4]
```diff
@@ -481,9 +705,18 @@ function mockConsideration(howMany: number = 10): ReceivedItemStruct[] {
   return consideration;
 }
 
+function mockBulkOrderHashes(howMany: number = 10): string[] {
+  const hashes = [];
+  for (let i = 0; i < howMany; i++) {
+    hashes.push(keccak256(`0x123${i >= 10 ? i + "0" : i}`));
```
Comment (frankisawesome) [inline test/immutablesignedzone.spec.ts:711]: This is whacky thing and only works with numbers 0-99..

**Review by lfportal — APPROVED** [review]

**Pull request merged**